model/softfloat/riscv_softfloat_eq.sail

```diff
@@ -0,0 +1,153 @@
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''                   */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED                    */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A                      */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR                  */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,                         */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT                     */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF                     */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND                  */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,                   */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT                   */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF                   */
+/*  SUCH DAMAGE.                                                                         */
+/*=======================================================================================*/
+
+/* **************************************************************** */
+/* This file implements the floating-point for equal and consumes   */
+/* by the softfloat API contract, riscv_softfloat_interface.sail.   */
+/*                                                                  */
+/* The below floating-point types are supported.                    */
+/* 1. Double-precision, aka 64 bits floating-point.                 */
+/* 2. Single-precision, aka 32 bits floating-point.                 */
+/* 3. Half-precision, aka 16 bits floating-point.                   */
+/*                                                                  */
+/* However, the implementation still need some global variables in  */
+/* file riscv_softfloat_interface.sail. It includes:                */
+/* 1. float_result.                                                 */
+/* 2. float_fflags.                                                 */
+/* 3. float_roundingMode.                                           */
+/* **************************************************************** */
+
+val            operator == : (bits_S, bits_S) -> bool
+val            operator |  : (bool, bool) -> bool
+val            operator |  : (bits_S, bits_S) -> bits_S
+val            operator << : (bits_S, int) -> bits_S
+val  f_is_at_least_one_NaN : forall 'm, 'm in {16, 32, 64}. (bits('m), bits('m)) -> bool
+val f_is_at_least_one_SNaN : forall 'm, 'm in {16, 32, 64}. (bits('m), bits('m)) -> bool
+
+val      riscv_softfloat_f16_eq : (bits_H, bits_H) -> unit
+function riscv_softfloat_f16_eq (op1, op2) = {
+  riscv_softfloat_enter (0b000);
+
+  var result : bits_H = zeros(16);
+
+  if f_is_at_least_one_NaN (op1, op2) then {
+    if f_is_at_least_one_SNaN (op1, op2) then {
```

Comment:
You can use `&` rather than nested ifs.

Also I think it's better to use functional style where it works nicely:

```
let invalid = (f_is_NaN (op1) | f_is_NaN(op2)) & (f_is_SNaN (op1) | f_is_SNaN(op2))
if invalid then riscv_softfloat_raise_flags (softfloat_flag_invalid);

let both_zero = (op1 | op2) << 1) == zeros();
let equal = not(invalid) & (op1 == op2 | both_zero);

riscv_softfloat_leave(zero_extend(bool_bits(equal))
```

I think that's a lot clearer.